### ISSUE TITLE
No xxx.zip provided when trying to deploy package using CodePipeline/CodeBuild

### ISSUE THREAD

**Issue by ghost** [serverless/serverless#9015]
I'm trying to deploy a serverless app with two functions using CodeBuild. In the serverless.yml I used settings for packaging individually with includes and excludes. When I remove these settings from the yaml file, everything runs perfectly. Because we are going to add a lot more functions in the future we really need this functionality to work with AWS CodeBuild.

I feel like this is somehow related to this bug which is closed recently: https://github.com/serverless/serverless/issues/4720

<details>
<summary><code>serverless.yml</code></summary>

```yaml
service: xxx
app: xxx-app

provider:
  name: aws
  runtime: nodejs12.x
  timeout: 10
  memorySize: 1546
  stage: ${opt:stage}
  iamRoleStatements:
      - Effect: 'Allow'
        Action:
          - 'secretsmanager:GetSecretValue'
        Resource:
              - ${self:custom.secretArn.${self:provider.stage}}
  region: eu-west-1
  apiGateway:
    restApiId: ${ssm:/api-gateway/${opt:apiName}/${self:provider.stage}/api-id~true}
    restApiRootResourceId: ${ssm:/api-gateway/${opt:apiName}/${self:provider.stage}/root-resource-id~true}

custom:
  secretArn:
    dev: xxx
  apiGatewayThrottling:
    maxRequestsPerSecond: 1000
    maxConcurrentRequests: 500
  apiKeys:
    - name: xxx
      value:
        encrypted: ${ssm:/kms/${opt:apiName}/${self:provider.stage}/api-key~true}
        kmsKeyRegion: eu-west-1

package:
  individually: true
  exclude:
    - ./**

functions:
  function1:
    handler: src/main/function1.getData
    package:
      include:
        - src/main/function1.js
        - src/libs/helper/**
        - node_modules/**
    events:
        - http:
            path: /x/x/x
            method: GET
            private: true
  function2:
    handler: src/main/function2.getData
    package:
      include:
        - src/main/function2.js
        - node_modules/**
    events:
        - http:
            path: /y/y/y
            method: GET
            cors: true

plugins:
  - serverless-api-gateway-throttling
  - serverless-add-api-key
  - serverless-aws-documentation
```

</details>

<details>
<summary><b><code>serverless deploy --stage $ENV --package $CODEBUILD_SRC_DIR/artifacts/$ENV -v</code> output</b></summary>

```

Serverless Error ---------------------------------------
--
  
 No function1.zip file found in the package path you provided.
 
```

</details>

<b>Installed version</b>

```
 Operating System:          linux
 Node Version:              12.16.1
 Framework Version:         2.27.1
 Plugin Version:            4.4.3
 SDK Version:               2.3.2
 Components Version:        3.7.1
```

**Comment by pgrzesik** [comment]
Hello @VincentHuysmans1, thanks for reporting. I've tried to reproduce your issue locally with a simplified version of your config but with no luck - in the approach you're using, you have to first package your functions with `package` command - could you share how are you doing that? Could you also list the contents of the directory `$CODEBUILD_SRC_DIR/artifacts/$ENV`? Are there function artifacts as well as files with CF templates?

**Comment by VincentHuysmans1** [comment]
The package command:

`serverless package --package artifacts/${ENV} --stage ${ENV}  -v`

The content of artifacts folder:

```
-rw-r--r-- 1 root root     2077 Feb 23 21:09 cloudformation-template-create-stack.json
--
31 | -rw-r--r-- 1 root root    18533 Feb 23 21:09 cloudformation-template-update-stack.json
32 | -rw-r--r-- 1 root root    33995 Feb 23 21:09 serverless-state.json
33 | -rw-r--r-- 1 root root 16918705 Feb 23 21:09 function1.zip
34 | -rw-r--r-- 1 root root 16916396 Feb 23 21:09 function2.zip

```
Thanks

**Comment by VincentHuysmans1** [comment]
@pgrzesik Do you want me to provide some additional information?

**Comment by pgrzesik** [comment]
Hello @VincentHuysmans1 - I was just looking into it and trying to reproduce your issue but didn't manage to as of yet. 

The listing of `artifacts` folder is from the separate build step or it's listed right before the `serverless deploy --stage $ENV --package $CODEBUILD_SRC_DIR/artifacts/$ENV -v` command is executed? It seems like the artifacts are not available inside the `$CODEBUILD_SRC_DIR/artifacts/$ENV` at the time the command gets executed

**Comment by VincentHuysmans1** [comment]
It is listed right before the deploy statement. When I remove the individually packaging from the serverless.yml, everything works just fine.

**Comment by pgrzesik** [comment]
While searching for potential clues, I've found a similar issue that a few people experienced in the past as well - https://github.com/serverless/serverless/issues/6964 - unfortunately, it seems like the fix didn't make it. Could you try using one of the workarounds proposed that uses relative path instead of the absolute one with `CODEBUILD_SRC_DIR`? 

@medikoo As I see that you've been involved with the original issue - do you happen to remember what was  the cause of it?

**Comment by medikoo** [comment]
> @medikoo As I see that you've been involved with the original issue - do you happen to remember what was the cause of it?

@pgrzesik in #6964 bug was incorrect handling of `--package` option. I remember then that I investigated it, and that handling of that setting was spread across many places, and I think it was doubled at some point (but I can't be sure now, what speficically I saw)

I started working then on refactor https://github.com/serverless/serverless/pull/7009, and while refactoring handling of this option was relatively straightforward. I observed tens of tests being broken in many files, which drive me to conclusion that first those tests need to be refactored to `runServerless`, and unfortunately it got stuck then.

On revisiting, I would first confirm whether same issue won't be fixed by solution proposed here: https://github.com/serverless/serverless/issues/7298 If it's the case, I would focus just on that.

Otherwise I would revisit failing tests, and first ensure we have them moved to `runServerless` variant, and then refactor handling of this option.

**Comment by pgrzesik** [comment]
Thank you for insights @medikoo :bow: I will dive a bit deep on that issue and see what I can find

**Comment by pgrzesik** [comment]
Hello @VincentHuysmans1 - thanks for patience here - unfortunatelly due to other priorities I won't be able to focus more on researching this issue during this week for sure - sorry about that. Did you try the proposed workaround from linked issue? Would that be an acceptable solution in the meantime?

**Comment by VincentHuysmans1** [comment]
Hi @pgrzesik - you too thanks for the time you've already spend on investigating. I have not tested the workaround yet, but I will try it today and come back at it.

**Comment by VincentHuysmans1** [comment]
@pgrzesik So I've tried using the relative path as described in #6964 - unfortunately I get the same error.

`serverless deploy --stage $ENV --package ./artifacts/$ENV -v`

**Comment by pgrzesik** [comment]
Thanks @VincentHuysmans1 - too bad that the workaround didn't help in this case - I'll try to dive deeper into that issue hopefully next week and report back with findings.

**Comment by ghostjester** [comment]
Do we have any news about this issue?

**Comment by pgrzesik** [comment]
Hello @ghostjester - unfortunately, due to other priorities I wasn't yet able to investigate it further.

**Comment by pgrzesik** [comment]
Hello @VincentHuysmans1 and @ghostjester - I've managed to find some time to investigate further but when using your specific example @VincentHuysmans1 I didn't manage to reproduce this issue locally. Are you able to deploy if you do following:
```
serverless package --package path/to/package
serverless deploy --package path/to/package
``` 
? 

I'm trying to exclude any issues related with how CODEBUILD extracts the cached artifacts, so that information would be really helpful.

Also, thanks a lot for patience, unfortunately a lot of other priorities came up which didn't let me to focus on this one.

**Comment by jplandry908** [comment]
Hi @pgrzesik. I am experiencing the same issue and came across this looking for a solution. I am using AWS CodePipeline to Build and then Deploy the solution. It is breaking up the packages and building correctly, but the Deploy process is failing - stating that the .zip file is not there for the individual packages, but I can confirm it is in that directory. I have traced it to an invalid path in the serverless-state.json file, and attempting to do a sed command to change the path to what it needs to be. I'll let you know if I can find a workaround, which may help implement a fix.

**Comment by jplandry908** [comment]
For anyone else experiencing this issue, I finally identified the root cause and a simple workaround.

In summary, the **"artifact"** incorrectly has **"$CODEBUILD_SRC_DIR"** in the value. Also, the _srcxxxxxxxxxx_ number changes between deploy and build, so you can't just do a find/replace for the exact string in DEPLOY since the value is the $CODEBUILD_SRC_DIR that was in BUILD.

For each function, you will see an artifact property that looks similar to:
`"artifact": "/codebuild/output/src887072543/src/.serverless/function.zip"`
while it should actually look like this:
`"artifact": ".serverless/function.zip"`

To work around this issue, I simply added the following two lines in my BUILD commands, immediately above the '**serverless deploy ...**' statement.

`- sed -i 's/\/codebuild\/output//g' artifacts/${!stage}/serverless-state.json`
`- sed -i 's/\/src.*\//.serverless\//g' artifacts/${!stage}/serverless-state.json`

This will transform the "/codebuild/output/src887072543/src/.serverless/function.zip" statements to ".serverless/function.zip", allowing the deploy process to find the file without error.

This took me a while to figure out, so I hope it helps others that may be experiencing the same issue.

**Comment by pessrini** [comment]
Hi @pgrzesik  @jplandry908 Im facing the same issue. I am using azure pipeline to package and deploy artifacts, packaging and deploying in different azure pipeline agents so its getting this issue.

For each function in the serverless-state.json, I can see the the artifact that look similar to:
`"artifact": "/root/agent09/_work/12/s/.serverless/service.zip"`
while deploy artifacts , this path will change depend on available pipeline agent like `"artifact": "/root/agent109/_work/12/s/.serverless/service.zip"`

**Installed Version**

`Framework Core: 2.72.1 (local)
Plugin: 5.5.4
SDK: 4.3.0
Components: 3.18.2
`

**Comment by jplandry908** [comment]
Hi @pessrini. If you are having the same issue that I was having, then you need to strip out the `/root/agent*` code from the artifact line.

So in your example, you have:
`"artifact": "/root/agent09/_work/12/s/.serverless/service.zip"`

And it should look like:
`"artifact": ".serverless/service.zip"`

Using the sed command, you can remove all text from two strings using the following command:
`sed 's/{START}.*{END}//'`

So instead of the command I specified above, for your situation try:
`sed -i 's/\/root\/.*\/s\///' artifacts/${!stage}/serverless-state.json`

The `-i` command says to change it in place, and then change `artifacts/${!stage}/serverless-state.json` to the path of your `serverless-state.json` file is located. Hopefully this works for you.

**Comment by czubocha** [comment]
Hey everyone, I just wanted to let you know that this issue should now be resolved in [v4.4.17](https://github.com/serverless/serverless/releases/tag/v4.4.17). If you encounter any further issues, feel free to reach out!
